File: app/src/main/java/com/example/popvm/services/VmController.kt
package com.example.popvm.services

import android.os.Handler
import android.os.Looper
import com.example.popvm.models.QemuConfigOptions
import com.example.popvm.models.TerminalLine
import com.example.popvm.models.VmStatus

class VmController(
    private var onStatusChange: ((VmStatus) -> Unit)? = null,
    private var onOutput: ((TerminalLine) -> Unit)? = null
) {
    private var qemu: QemuProcess? = null
    private var status: VmStatus = VmStatus.STOPPED
    private var config: QemuConfigOptions = QemuConfig.getDefaultOptions()
    private val handler = Handler(Looper.getMainLooper())

    fun setCallbacks(onStatusChange: (VmStatus) -> Unit, onOutput: (TerminalLine) -> Unit) {
        this.onStatusChange = onStatusChange
        this.onOutput = onOutput
    }

    fun setConfig(options: QemuConfigOptions) {
        this.config = options
    }

    fun getConfig(): QemuConfigOptions = config

    fun getStatus(): VmStatus = status

    fun start() {
        if (qemu != null) return

        setStatus(VmStatus.STARTING)

        qemu = QemuProcess(
            onOutput = { line ->
                onOutput?.invoke(line)
            },
            onStateChange = { running ->
                if (!running) {
                    setStatus(VmStatus.STOPPED)
                    qemu = null
                }
            }
        )

        val args = QemuConfig.buildArgs(config)
        qemu?.start(args, config.memoryMb, config.smpCores)

        handler.postDelayed({
            if (status == VmStatus.STARTING) {
                setStatus(VmStatus.RUNNING)
            }
        }, 4500)
    }

    fun stop() {
        if (qemu == null) return
        setStatus(VmStatus.STOPPING)
        qemu?.stop()
        qemu = null
        setStatus(VmStatus.STOPPED)
    }

    fun sendCommand(cmd: String) {
        qemu?.executeCommand(cmd)
    }

    private fun setStatus(newStatus: VmStatus) {
        status = newStatus
        onStatusChange?.invoke(newStatus)
    }
}

File: src/services/VmController.ts
import { QemuConfigOptions, TerminalLine, VmStatus } from '../types';
import { QemuProcess } from './QemuProcess';
import { QemuConfig } from './QemuConfig';

export class VmController {
  private qemu: QemuProcess | null = null;
  private status: VmStatus = 'STOPPED';
  private config: QemuConfigOptions = QemuConfig.getDefaultOptions();

  constructor(
    private onStatusChange: (status: VmStatus) => void,
    private onOutput: (line: TerminalLine) => void
  ) {}

  setConfig(options: QemuConfigOptions): void {
    this.config = options;
  }

  getConfig(): QemuConfigOptions {
    return this.config;
  }

  getStatus(): VmStatus {
    return this.status;
  }

  start(): void {
    if (this.qemu) return;

    this.setStatus('STARTING');

    this.qemu = new QemuProcess(
      (line) => this.onOutput(line),
      (running) => {
        if (!running) {
          this.setStatus('STOPPED');
          this.qemu = null;
        }
      }
    );

    this.qemu.start(
      this.config.distroName,
      this.config.memoryMb,
      this.config.smpCores,
      this.config.diskSizeGb,
      this.config.isoFile
    );

    setTimeout(() => {
      if (this.status === 'STARTING') {
        this.setStatus('RUNNING');
      }
    }, 3400);
  }

  stop(): void {
    if (!this.qemu) return;
    this.setStatus('STOPPING');
    this.qemu.stop();
    this.qemu = null;
    this.setStatus('STOPPED');
  }

  sendCommand(cmd: string): void {
    this.qemu?.executeCommand(cmd);
  }

  private setStatus(newStatus: VmStatus): void {
    this.status = newStatus;
    this.onStatusChange(newStatus);
  }
}
